sdb/commands/nvpair/nvpair.py

```diff
@@ -0,0 +1,209 @@
+            self.get_array_values(type_)
+            return
+        if "nvlist" in type_:
+            print("values=")
+            self.nvl.set_indent(self.nvl.indent + 4)
+            self.nvl.print_one(sdb.create_object("nvlist_t *", self.data_ptr))
+            self.nvl.set_indent(self.nvl.indent - 4)
+            return
+        print("value=", end='')
+        if "boolean_value" in type_:
+            type_ = "uint32_t"
+        elif "boolean" in type_:  # DATA_TYPE_BOOLEAN has name and no value
+            print("")
+        elif "hrtime" in type_:
+            type_ = "int64_t"
+        if "int" in type_:
+            print(self.get_value_int(type_, self.data_ptr))
+        if "string" in type_:
+            print(
+                sdb.create_object("char *",
+                                  self.data_ptr).string_().decode("utf-8"))
+        if "byte" in type_:
+            print(hex(int(self.get_value_int("unsigned char", self.data_ptr))))
+        return
+
+
+class Nvlist(sdb.SingleInputCommand):
+    """
+    Print nvlist_t
+
+    DESCRIPTION
+        Print all the nvpair_t in the passed nvlist_t.  Handle basic types,
+        array types, and nvlist_t types.  Type double is omitted as it is
+        not used in zfs.
+
+    EXAMPLE
+    Print nvlist_t of snapshot holds from the nvlist_t * pointer address:
+
+        sdb> echo 0xffff970d0ff681e0 | nvlist
+        name=monday-1 type=DATA_TYPE_UINT64 value=1633989858
+        name=monday-2 type=DATA_TYPE_UINT64 value=1633989863
+    """
+
+    names = ["nvlist"]
+    input_type = "nvlist_t *"
+    output_type = "nvlist_t *"
+    indent = 0
+
+    def set_indent(self, indent: int) -> None:
+        self.indent = indent
+
+    # nvlist iteration methods
+    @staticmethod
+    def nvlist_contains_nvp(nvl: drgn.Object, nvp: drgn.Object) -> int:
+        priv = drgn.cast("nvpriv_t *", nvl.nvl_priv)
+
+        if nvp.address_ == 0:
+            return 0
+
+        curr = priv.nvp_list
+        while not sdb.is_null(curr):
+            if curr.nvi_nvp.address_ == nvp.address_:
+                return 1
+            # pylint: disable=protected-access
+            curr = curr._nvi_un._nvi._nvi_next
+
+        return 0
+
+    @staticmethod
+    def nvlist_first_nvpair(nvl: drgn.Object) -> drgn.Object:
+        if sdb.is_null(nvl) or sdb.is_null(nvl.nvl_priv):
+            return None
+        priv = drgn.cast("nvpriv_t *", nvl.nvl_priv)
+        return priv.nvp_list.nvi_nvp
+
+    @staticmethod
+    def nvlist_next_nvpair(nvl: drgn.Object, nvp: drgn.Object) -> drgn.Object:
+        if sdb.is_null(nvl) or sdb.is_null(nvl.nvl_priv):
+            return None
+
+        priv = drgn.cast("nvpriv_t *", nvl.nvl_priv)
+
+        curr_addr = nvp.address_ - drgn.offsetof(sdb.get_type("i_nvp_t"),
+                                                 "nvi_nvp")
+        curr = sdb.create_object("i_nvp_t *", curr_addr)
+
+        if priv.nvp_curr == curr or Nvlist.nvlist_contains_nvp(nvl, nvp):
```

Comment:
Is the nvlist_contains_nvp() check really necessary?  Seems like it would make overall iteration O(N^2) since every call to next_nvpair() will iterate the whole nvlist.  Maybe this isn't a big deal as long as your list isn't TOO long (<1000?)

Reply:
Good question.  I basically ported
https://src.illumos.org/source/xref/illumos-gate/usr/src/common/nvpair/nvpair.c?r=b8a5bee1#1431
Let play around and see if or how often nvlist_contains_nvp is needed.

Reply:
I think the extra check is there in case the nvlist is having element(s) removed while we're traversing it.